### PR TITLE
Lint to check alphabetical order of uses, mods & crates

### LINKED ISSUE #7451: Hand over the alphabetical checking to lints!
If you have a look at #7441, #7390, and #7112, the alphabetical checking for `extern crate`, `use` and `mod` statements are done by tidy. @SimonSapin just told me that all those could be done better with Rust's lints. We already have some lints written by dear @Manishearth, but we need more!

The future code will live at `components/plugins/lints` along with the other lints.

### PR DESCRIPTION
**Pull request by ghost**
fixes #7451

[Review on Reviewable](https://reviewable.io/reviews/servo/servo/7546)

### CODE REVIEW

**Comment by Wafflespeanut** [comment]
NOTE: This is just a temporary commit!

**Comment by nox** [comment]
This has the exact same problem as the Python lint, in which import lists aren't checked for sortedness, right?

I mean things like:

```rust
use mod::{foo, bar, qux};
```

**Comment by Wafflespeanut** [comment]
@nox No, only the individual declarations. I think checking for sortedness in import lists is probably a bad idea, since some of the giant files have a lot of lists and arranging them would be pain :)

**Comment by Manishearth** [comment]
If the lint outputs a sorted list to be used, it wouldn't be so bad. But this PR is fine too, we can discuss whether or not to augment the functionality.

**Comment by Manishearth** [comment]
Reviewed 3 of 3 files at r1.
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

<sup>**[components/plugins/lints/sorter.rs, line 37 \[r1\]](https://reviewable.io:443/reviews/servo/servo/7546#-JydgrHhGholNtK7TdTK)** ([raw file](https://github.com/servo/servo/blob/2d6109f5aaf7801686d6deb9853a00fdd110b0ab/components/plugins/lints/sorter.rs#L37)):</sup>
We may want to glob import in the future, so blocking `std::prelude` is a better idea here. But this is okay too.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7546)

**Comment by Wafflespeanut** [comment]
Oh yeah, right. I now remember that we'll be actually suggesting how the stuff should be sorted :)

**Comment by Wafflespeanut** [comment]
One tiny thing left - checking for list imports...

**Comment by Wafflespeanut** [comment]
Now, there's a `InternedString` <--> `String` problem

**Comment by Manishearth** [comment]
You don't need to put r=Manishearth in your commits. Homu will do that in the merge commit

**Comment by Wafflespeanut** [comment]
I'm not sure whether it's pretty, but we have a working lint! ^_^

Here's how crates are checked (same goes for mods)...

crates

Here's how use lists are checked (cc @nox)...

use_list

**Comment by Wafflespeanut** [comment]
This still needs more polishing...

**Comment by jdm** [comment]
Calling the lint sorted_declarations means that `deny(sorted_declarations)` actually means "disallow unsorted declarations", which is a bit confusing.

**Comment by Wafflespeanut** [comment]
Yeah, right. That should be `deny(unsorted_declarations)` :)

**Comment by Manishearth** [comment]
Reviewed 1 of 1 files at r2, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

<sup>**[components/plugins/lints/sorter.rs, line 34 \[r2\]](https://reviewable.io:443/reviews/servo/servo/7546#-JzD0pyYN5VHfgXJ3PaF)** ([raw file](https://github.com/servo/servo/blob/f4f19e8af8442fc916760320991888e91cdf97e0/components/plugins/lints/sorter.rs#L34)):</sup>
Also ignore mods with attributes

---

<sup>**[components/plugins/lints/sorter.rs, line 55 \[r2\]](https://reviewable.io:443/reviews/servo/servo/7546#-JzD1d2hj6km-CmMd-Oc)** ([raw file](https://github.com/servo/servo/blob/f4f19e8af8442fc916760320991888e91cdf97e0/components/plugins/lints/sorter.rs#L55)):</sup>
You need to sort it so that `self` comes at the beginning of the list

---

<sup>**[components/plugins/lints/sorter.rs, line 87 \[r3\]](https://reviewable.io:443/reviews/servo/servo/7546#-JzHHnvLfxwQ3Ii7TZDF)** ([raw file](https://github.com/servo/servo/blob/dda080496a4bc88e7410af5bbfee39cf2cdaa8a6/components/plugins/lints/sorter.rs#L87)):</sup>
&Vec should be &[..]


Also, `T: Deref<Target=String>` should be enough, reallyx

---

<sup>**[components/plugins/lints/sorter.rs, line 93 \[r2\]](https://reviewable.io:443/reviews/servo/servo/7546#-JzD2VaXx45B0yE8H26B)** ([raw file](https://github.com/servo/servo/blob/f4f19e8af8442fc916760320991888e91cdf97e0/components/plugins/lints/sorter.rs#L93)):</sup>
Error message should be better.

Something like

{} should be in alphabetical order\n\texpected: {}\n\tfound: {}

where the first {} is either "use statements", "module declarations", or "crate declarations"

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7546)

**Comment by Wafflespeanut** [comment]
This now shows something like this...

screenshot

**Comment by Wafflespeanut** [comment]
Now, all that's left is sort the declarations in the existing files (which is gonna be painful!)

File: components/plugins/lints/sorter.rs [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use rustc::lint::{Context, LintPass, LintArray};
+use std::cmp::Ordering;
+use std::fmt::Display;
+use syntax::ast::{Mod, Item_, PathListItem_, ViewPath_};
+use syntax::codemap::{Pos, BytePos, Span};
+use syntax::print::pprust::path_to_string;
+
+declare_lint!(UNSORTED_DECLARATIONS, Deny,
```
Comment (Manishearth) [inline components/plugins/lints/sorter.rs]: This should be `Warn`. `Deny` is going to be very annoying when you're fiddling with imports temporarily before checking in.

File: components/plugins/lints/sorter.rs [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+                                    true => Ordering::Less,
+                                    false => a.cmp(b),
+                                }
+                            });
+                            for i in 0..old_list.len() {
+                                if old_list[i] != new_list[i] {
+                                    let suggestion = format!("use lists should be in alphabetical order!\
+                                                              \n\texpected order of list: {{{}}}\
+                                                              \n\tlist found: {{{}}}",
+                                                              new_list.join(", "), old_list.join(", "));
```
Comment (Manishearth) [inline components/plugins/lints/sorter.rs]: Not like this. 

```
Use lists should be in alphabetical order.
try:
use foo::bar::baz::{a,b,c,d};
```

print out a fully copyable use statement

File: components/plugins/lints/sorter.rs [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+                                .map(|&(ref string, _span)| string.clone())
+                                .collect::<Vec<T>>();
+            new_slice.sort();
+            for i in 0..old_slice.len() {
+                let (name, mut span) = (old_slice[i].0.clone(), old_slice[i].1);
+                span.lo = span.lo + BytePos::from_usize(syntax_len + 1);
+                if name != new_slice[i] {
+                    let suggestion = format!("{} should be in alphabetical order!\n\t\
+                                              expected: {}\n\tfound: {}",
+                                              kind, new_slice[i], name);
```
Comment (Manishearth) [inline components/plugins/lints/sorter.rs]: ditto.

**Comment by Wafflespeanut** [comment]
This is still incomplete. Firstly, I gotta cripple the lint, so that it doesn't walk into `servo/target/...` and check the generated DOM bindings. And second, I haven't yet given any preference to the `pub` stuff. So far, the sorting is based on the names of the declarations.

**Comment by jdm** [comment]
Alternatively, we could fix the generated bindings.

**Comment by Wafflespeanut** [comment]
This just keeps getting better and better! Now, we gotta filter the additional attributes that come clinging with the mod declarations (for instance, [this](https://mxr.mozilla.org/servo/source/components/style/lib.rs#53) shows `#[allow(non_camel_case_types)]` even though it's not actually declared there, but [here](https://mxr.mozilla.org/servo/source/components/style/values.rs#5)).

Also, the lint doesn't care about renamed declarations (for now). So, things like
``` rust
extern crate foobar as foo
```
are just interpreted as `extern crate foo`, which should also be fixed as well.

**Comment by Wafflespeanut** [comment]
(current status)

screenshot

**Comment by Wafflespeanut** [comment]
Alright, tidy doesn't give up. It expects `pub mod` before mod, but our sorting is something like `#[macro_use]` at the start, followed by mods and then followed by `pub mod`s.

**Comment by jdm** [comment]
It seems to me that the main downside of doing this check in a lint vs. in the python script is that we lose the immediate feedback that travis provides, so we won't notice that a PR needs amending until it fails to merge.

**Comment by Wafflespeanut** [comment]
@jdm Agreed. Also, since the lint only throws warnings, we can't know whether a PR has merged with warnings :/

@Manishearth maybe we should go back to throwing errors! :)

**Comment by Manishearth** [comment]
No. No errors. Go does this with unused imports, and it's _extremely_ annoying when this happens while you're developing. So much that a myriad of solutions, ranging from auto-import-fixers to modified compilers have come out.

Style lints should not be errors.

**Comment by Wafflespeanut** [comment]
Okay, I've planned to move this lint to [rustfmt](https://github.com/nrc/rustfmt/issues/298). I'll just submit a PR for the files (well, I can't just abandon it, it's consumed a lot of time! :P)

**Comment by Wafflespeanut** [comment]
Okay, I've planned to move this lint elsewhere. I'll just submit a PR for the files (well, I can't just abandon it, it's consumed a lot of time! :P)

**Pull request closed without merging**